openvdb_ax/openvdb_ax/ax.h

```diff
@@ -0,0 +1,95 @@
+// Copyright Contributors to the OpenVDB Project
+// SPDX-License-Identifier: MPL-2.0
+
+/// @file ax.h
+///
+/// @author Nick Avramoussis
+///
+/// @brief  Single header include which provides methods for initializing AX and
+///   running a full AX pipeline (pasrsing, compiling and executing) across
+///   standard OpenVDB Grid types.
+///
+/// @details  These methods wrap the internal components of OpenVDB AX to
+///   provide easier and quick access to running AX code. Users who wish to
+///   further optimise and customise the process may interface with these
+///   components directly. See the body of the methods provided in this file for
+///   example implementations.
+
+#ifndef OPENVDB_AX_AX_HAS_BEEN_INCLUDED
+#define OPENVDB_AX_AX_HAS_BEEN_INCLUDED
+
+#include <openvdb/openvdb.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace ax {
+
+/// @brief  Initializes OpenVDB AX and subsequent LLVM components.
+/// @details  Must be called before any AX compilation or execution is performed.
+///           Can be safely called from multiple threads. Cannot be called after
+///           uninitialize has been called.
+/// @note   The implementation for the initialize methods can be found in
+///         compiler/Compiler.cc
+void initialize();
+
+/// @brief  Check to see if OpenVDB AX components have been initialized.
+/// @note  Can be safely called from multiple threads.
+bool isInitialized();
+
+/// @brief  Uninitialize and deregister OpenVDB AX.
+/// @details  This has the important function of shutting down LLVM and
+///           correctly freeing statically allocated LLVM types. Should be
+///           called on application termination. Can be safely called from
+///           multiple threads.
+void uninitialize();
+
+////////////////////////////////////////
+////////////////////////////////////////
+
+/// @brief  Run a full AX pipeline (parse, compile and execute) on a single
+///         OpenVDB Grid.
+/// @details  This method wraps the parsing, compilation and execution of AX
+///           code for a single OpenVDB grid of any standard grid type
+///           (including OpenVDB Points Grids). Provided AX code is expected to
+///           only refer to the provided single grid. On success, the grid will
+///           have its voxels or point data modified as dictated by the provided
+///           AX code.
+/// @note  Various defaults are applied to this pipeline to provide a simple
+///        run signature. For OpenVDB Numerical grids, only active voxels are
+///        processed. For OpenVDB Points grids, all points are processed. Any
+///        warnings generated by the parser, compiler or executable will be
+///        ignored.
+/// @note  Various runtime errors may be thrown from the different AX pipeline
+///        stages. See Exceptions.h for the possible different errors.
```

Comment:
I've incorporated the logging changes (see `compiler/Logger.h`). These run functions haven't been updated but will add methods in the future which take a `Logger` here too. On the compiler, you can now call `exe = C.compiler<T>("@a = 1;", &logger);`. This will return a nullptr on failure and update the logger accordingly. Exceptions have been limited to those which cause unexpected behaviour (incorrect dev api usage) and can probably be further reduced to asserts in the future.

Reply:
Looking forward to the logger based run() method in future PRs.